### PR TITLE
Recovery remove new

### PR DESCRIPTION
**Pull request by abudnik**
This PR depends on PR #120.

### CODE REVIEW

File: recovery/elliptics_recovery/types/dc.py [hunk 1]
```diff
@@ -234,18 +250,84 @@ def contains(addresses_with_backends, address, backend_id):
...
+
+                    failed_tasks = []
+                    is_last_attempt = attempt == ctx.attempts - 1
+                    for i, r in enumerate(responses):
+                        key, group_id = tasks[i]
+                        status = r.get()[0].status
+                        log.info('Removed uncommitted key: %s, group: %s, status: %s, last attempt: %s',
+                                 key, group_id, status, is_last_attempt)
+                        statuses[(key, group_id)] = status
+                        if status in (-errno.ETIMEDOUT, -errno.ENXIO):
```
Comment (shaitan) [inline recovery/elliptics_recovery/types/dc.py]: Why does it count only these two errors as a failure which can be overcome after retry? Let's invert it to `status not in (0, -errno.ENOENT, -errno.EBADFD)`?

File: recovery/elliptics_recovery/dc_recovery.py [hunk 2]
```diff
@@ -50,34 +51,38 @@ def remove(self):
...
+
+        self.on_complete()
 
     def on_remove_corrupted(self, results, error):
         try:
+            for r in results:
+                self.results[r.group_id] = r.status
+
             if error.code:
                 self.stats_cmd.counter('remove.{0}'.format(error.code), 1)
```
Comment (abudnik) [inline recovery/elliptics_recovery/dc_recovery.py]: fixed

File: recovery/elliptics_recovery/dc_recovery.py [hunk 3]
```diff
@@ -50,34 +51,38 @@ def remove(self):
...
                 self.stats.remove_failed += 1
                 failed_groups = [r.group_id for r in results if r.status != 0]
```
Comment (abudnik) [inline recovery/elliptics_recovery/dc_recovery.py]: fixed

File: recovery/elliptics_recovery/dc_recovery.py [hunk 4]
```diff
@@ -429,40 +438,16 @@ def onwrite(self, results, error):
...
             self.stop(False)
 
-    def onremove(self, results, error):
-        try:
-            if error.code:
-                self.stats_cmd.counter('remove.{0}'.format(error.code), 1)
-                self.stats.remove_failed += 1
-                failed_groups = [r.group_id for r in results if r.status != 0]
-                # NB: used hard-coded 'removed_uncommitted_keys', since remove method is called only for uncommitted keys
-                self.stats.removed_uncommitted_keys += len(self.remove_session.groups) - len(failed_groups)
```
Comment (abudnik) [inline recovery/elliptics_recovery/dc_recovery.py]: fixed: `removed_uncommitted_keys` statistics used in process_uncommitted()

File: recovery/elliptics_recovery/types/dc.py [hunk 5]
```diff
@@ -234,18 +250,84 @@ def contains(addresses_with_backends, address, backend_id):
     return address_range
 
 
-def final_merge(ctx, results):
+def process_uncommitted(ctx, results):
```
Comment (abudnik) [inline recovery/elliptics_recovery/types/dc.py]: fixed

File: recovery/elliptics_recovery/types/dc.py (hunk 1 above)
Comment (abudnik) [inline recovery/elliptics_recovery/types/dc.py]: done

File: recovery/elliptics_recovery/types/dc.py [hunk 6]
```diff
@@ -173,44 +170,61 @@ def merged_results(ctx, results):
...
+    def __init__(self, filename, uncommitted_filename, dump_filename, prepare_timeout, safe, dump_keys):
+        self.filename = filename
+        self.uncommitted_filename = uncommitted_filename
+        self.dump_filename = dump_filename
+        self.prepare_timeout = prepare_timeout
+        self.safe = safe
+        self.dump_keys = dump_keys
+        self.newest_key_stats = defaultdict(int)
+
+    def on_key_data(self, key_data, f, uf, df):
```
Comment (shaitan) [inline recovery/elliptics_recovery/types/dc.py]: Please, don't use names like `f`, `uf`, `df`, especially as arguments)

File: recovery/elliptics_recovery/dc_recovery.py [hunk 7]
```diff
@@ -50,34 +50,38 @@ def remove(self):
...
                 remove_result = self.session.remove(self.key)
                 remove_result.connect(self.on_remove_corrupted)
+                return
             except:
                 log.exception("Failed to remove key: {0} from groups: {1}".format(self.key, self.session.groups))
-                self.callback()
+
+        self.on_complete()
 
     def on_remove_corrupted(self, results, error):
```
Comment (shaitan) [inline recovery/elliptics_recovery/dc_recovery.py]: Just to be clear, there is no more corruption flavor here, so let's make it just `on_remove`, ok?

File: recovery/elliptics_recovery/dc_recovery.py [hunk 8]
```diff
@@ -135,35 +145,6 @@ def run(self):
...
-            # they will be corresponding as different and will be overwritten
-            same_infos = [info for info in same_infos if info not in same_uncommitted]
-            incomplete_groups = [info.group_id for info in same_uncommitted]
-            same_groups = [info.group_id for info in same_infos]
-            log.info('Key: {0} has uncommitted replicas in groups: {1} and completed replicas in groups: {2}.'
-                     'The key will be recovered at groups with uncommitted replicas too.'
-                     .format(self.key, incomplete_groups, same_groups))
-
         same_meta = lambda lhs, rhs: (lhs.timestamp, lhs.size, lhs.user_flags) == (rhs.timestamp, rhs.size, rhs.user_flags)
         same_infos = [info for info in self.key_infos if same_meta(info, same_infos[0])]
```
Comment (shaitan) [inline recovery/elliptics_recovery/dc_recovery.py]: [Previous](https://github.com/interiorem/elliptics/pull/121/files?diff=unified#diff-1ad88f87dd8a16b8e7931edbce6921ebR146) filtering can be omitted, isn't it?

File: recovery/elliptics_recovery/types/dc.py [hunk 9]
```diff
@@ -171,30 +170,61 @@ def merged_results(ctx, results):
...
-                counter += 1
-                dump_key_data(key_data, f)
-
-                key_infos = key_data[1]
-                newest_key_group = key_infos[0].group_id
-                newest_key_stats[newest_key_group] = newest_key_stats.get(newest_key_group, 0) + 1
-
-                if ctx.dump_keys:
-                    df.write('{0}\n'.format(key_data[0]))
+    with open(filename, 'w') as f, open(uncommitted_filename, 'w') as uf, open(dump_filename, 'w') as df:
```
Comment (shaitan) [inline recovery/elliptics_recovery/types/dc.py]: Do these files really needed outside of `MergedKeys`?

File: recovery/elliptics_recovery/types/dc.py [hunk 10]
```diff
@@ -234,18 +264,107 @@ def contains(addresses_with_backends, address, backend_id):
...
+    Removes uncommitted keys. If a key has any committed replicas, then this key is
+    appended to the file containing committed keys.
+    If uncommitted key's replica hasn't exceeded prepare timeout, then skip recovering of the key,
+    because the key is under writing and can be committed in the nearest future.
+    '''
+    if ctx.dry_run or ctx.safe:
+        return
+
+    elog = elliptics.Logger(ctx.log_file, int(ctx.log_level))
+    node = elliptics_create_node(address=ctx.address,
```
Comment (shaitan) [inline recovery/elliptics_recovery/types/dc.py]: Why isn't `elliptics.create_node()` used?

File: recovery/elliptics_recovery/types/dc.py [hunk 11]
```diff
@@ -234,18 +264,107 @@ def contains(addresses_with_backends, address, backend_id):
...
+                        session.timeout = timeouts[group_id]
+                        responses.append(session.remove(key))
+
+                    failed_tasks = []
+                    is_last_attempt = attempt == ctx.attempts - 1
+                    for i, r in enumerate(responses):
+                        key, group_id, _ = tasks[i]
+                        status = r.get()[0].status
+                        log.info('Removed uncommitted key: %s, group: %s, status: %s, last attempt: %s',
+                                 key, group_id, status, is_last_attempt)
```
Comment (shaitan) [inline recovery/elliptics_recovery/types/dc.py]: `is_last_attempt` is less informative then `{attempt}/{attempts}`, can we use last variant?

Comment (shaitan) [inline recovery/elliptics_recovery/types/dc.py]: The same [here](https://github.com/interiorem/elliptics/pull/121/files?diff=unified#diff-cfba9cc0edcd75a4553b85a812857383R369).

File: recovery/elliptics_recovery/types/dc.py (hunk 9 above)
Comment (abudnik) [inline recovery/elliptics_recovery/types/dc.py]: Yes, all of them.

Comment (shaitan) [inline recovery/elliptics_recovery/types/dc.py]: Why are they needed **outside of** `MergedKeys`? Whey aren't they hidden inside `MergedKeys` object?

Comment (abudnik) [inline recovery/elliptics_recovery/types/dc.py]: All MergedKeys objects live until recovery completion. These files are needed only in this function (merge_results). And with a help of python `with` they are correctly released.

File: recovery/elliptics_recovery/types/dc.py (hunk 10 above)
Comment (abudnik) [inline recovery/elliptics_recovery/types/dc.py]: What is the difference?

File: recovery/elliptics_recovery/types/dc.py (hunk 9 above)
Comment (shaitan) [inline recovery/elliptics_recovery/types/dc.py]: Looks like you try to combine two entities inside `MergedKeys`, an active one which corresponds for filtering, collecting and counting keys' data and a passive one which travels across recovery. It leads to such awkward code. Please, split these two entities and hide excessive details from this code.

File: recovery/elliptics_recovery/types/dc.py (hunk 10 above)
Comment (shaitan) [inline recovery/elliptics_recovery/types/dc.py]: `elliptics_create_node()` is a legacy method from the time there was no `elliptics.create_node()`, so it shouldn't be used in new code.

File: recovery/elliptics_recovery/types/dc.py (hunk 9 above)
Comment (abudnik) [inline recovery/elliptics_recovery/types/dc.py]: No, I disagree. `MergedKeys` class seems pretty solid to me. Both `on_key_data` method and state of a `MergedKeys` are tightly interconnected.

>It leads to such awkward code.

This is a false conclusion. I can handle opened files in `MergedKeys` class. Ok?

File: recovery/elliptics_recovery/types/dc.py (hunk 6 above)
Comment (abudnik) [inline recovery/elliptics_recovery/types/dc.py]: fixed

File: recovery/elliptics_recovery/dc_recovery.py (hunk 7 above)
Comment (abudnik) [inline recovery/elliptics_recovery/dc_recovery.py]: fixed

File: recovery/elliptics_recovery/dc_recovery.py (hunk 8 above)
Comment (abudnik) [inline recovery/elliptics_recovery/dc_recovery.py]: fixed

File: recovery/elliptics_recovery/types/dc.py (hunk 10 above)
Comment (abudnik) [inline recovery/elliptics_recovery/types/dc.py]: fixed

File: recovery/elliptics_recovery/types/dc.py (hunk 11 above)
Comment (abudnik) [inline recovery/elliptics_recovery/types/dc.py]: fixed

File: recovery/elliptics_recovery/types/dc.py [hunk 12]
```diff
@@ -234,18 +265,105 @@ def contains(addresses_with_backends, address, backend_id):
...
+                statuses = {} # (key, group_id) -> status
+                for attempt in range(ctx.attempts):
+                    if not tasks:
+                        break
+
+                    batch_sizes = defaultdict(int) # group_id -> batch_size
+                    for _, group_id, key_size in tasks:
+                        batch_sizes[group_id] += key_size
+
+                    timeouts = {group_id: max(60, batch_size / ctx.data_flow_rate)
```
Comment (shaitan) [inline recovery/elliptics_recovery/types/dc.py]: Why are timeouts calculated in such way?

File: recovery/elliptics_recovery/types/dc.py [hunk 13]
```diff
@@ -234,18 +265,105 @@ def contains(addresses_with_backends, address, backend_id):
...
+                        if status not in (0, -errno.ENOENT, -errno.EBADFD):
+                            failed_tasks.append(tasks[i])
+                    tasks = failed_tasks
+
+                for key, key_infos in batch:
+                    infos = []
+                    for info in key_infos:
+                        if info.flags & elliptics.record_flags.uncommitted:
+                            status = statuses[(key, info.group_id)]
+                            if status == -errno.EBADFD:
```
Comment (shaitan) [inline recovery/elliptics_recovery/types/dc.py]: It still has to have an in-code comment that describes its logic.

File: recovery/elliptics_recovery/types/dc.py [hunk 14]
```diff
@@ -234,18 +265,105 @@ def contains(addresses_with_backends, address, backend_id):
...
+                                infos = None
+                                break
```
Comment (shaitan) [inline recovery/elliptics_recovery/types/dc.py]: It should also update skipped keys statistics if it excludes keys from recovering.

File: recovery/elliptics_recovery/types/dc.py [hunk 15]
```diff
@@ -234,18 +265,105 @@ def contains(addresses_with_backends, address, backend_id):
...
+                        statuses[(key, group_id)] = status
+
+                        if status == 0:
+                            stats.counter('removed_corrupted_keys', 1)
+                        else:
+                            stats_cmd.counter('remove.{0}'.format(status), 1)
+
+                        if status not in (0, -errno.ENOENT, -errno.EBADFD):
+                            failed_tasks.append(tasks[i])
+                    tasks = failed_tasks
```
Comment (shaitan) [inline recovery/elliptics_recovery/types/dc.py]: It should also update `remove_retries` statistics.

File: recovery/elliptics_recovery/types/dc.py (hunk 12 above)
Comment (abudnik) [inline recovery/elliptics_recovery/types/dc.py]: It was taken from mds proxy. It is also used in lrc tools and in dc_server_send.py.

File: recovery/elliptics_recovery/types/dc.py [hunk 16]
```diff
@@ -234,18 +265,105 @@ def contains(addresses_with_backends, address, backend_id):
...
+                    failed_tasks = []
+                    for i, r in enumerate(responses):
+                        key, group_id, _ = tasks[i]
+                        status = r.get()[0].status
+                        log.info('Removed uncommitted key: %s, group: %s, status: %s, attempts: %s/%s',
+                                 key, group_id, status, attempt, ctx.attempts)
+                        statuses[(key, group_id)] = status
+
+                        if status == 0:
+                            stats.counter('removed_corrupted_keys', 1)
```
Comment (abudnik) [inline recovery/elliptics_recovery/types/dc.py]: Sure, fixed.

File: recovery/elliptics_recovery/types/dc.py (hunk 13 above)
Comment (abudnik) [inline recovery/elliptics_recovery/types/dc.py]: Added comment

File: recovery/elliptics_recovery/types/dc.py (hunk 14 above)
Comment (abudnik) [inline recovery/elliptics_recovery/types/dc.py]: Added stat

File: recovery/elliptics_recovery/types/dc.py (hunk 15 above)
Comment (abudnik) [inline recovery/elliptics_recovery/types/dc.py]: Added stat

File: recovery/elliptics_recovery/types/dc.py (hunk 12 above)
Comment (abudnik) [inline recovery/elliptics_recovery/types/dc.py]: ... and in KeyRemover from dc_recovery.py

Comment (shaitan) [inline recovery/elliptics_recovery/types/dc.py]: Why does it use `batch_size`? Each key is removed by a single command, why do you specify timeout as for full batch?

Comment (abudnik) [inline recovery/elliptics_recovery/types/dc.py]: It's important to take into account that some keys of a batch can be large, so removal operation may take a significant amount of time: reading from disk, checksumming. Last keys of a batch will wait in a queue until first keys are being processed.

**Pull request merged**